ocular/src/chain/client/cache.rs

```diff
@@ -0,0 +1,519 @@
+pub struct Cache {
+    pub grpc_endpoint_cache: Box<dyn GrpcCache>,
+}
+
+/// Cache accessor defintions
+pub trait GrpcCache {
+    /// Check if cache has been initialized
+    fn is_initialized(&self) -> bool;
+    /// Add item to cache
+    fn add_item(&mut self, item: String) -> Result<(), CacheError>;
+    /// Remove item from cache
+    fn remove_item(&mut self, item: String) -> Result<(), CacheError>;
+    /// Retrieves a copy of all items from cache
+    fn get_all_items(&self) -> Result<HashSet<String>, CacheError>;
+}
+
+/// Cache initialization definitions
+impl Cache {
+    /// Constructor for file cache path. Must include filename with ".toml" suffix for file type.
+    ///
+    /// If override_if_exists is true it will wipe the previous file found if it exists.
+    /// Otherwise if set to false, it will use what is found at the file or create a new one if not found.
+    ///
+    /// Toml will be required to be structured as so:
+    /// [[endpoints]]
+    ///     address = "35.230.37.28:9090"
+    pub fn create_file_cache(
+        file_path: Option<&str>,
+        override_if_exists: bool,
+    ) -> Result<Cache, CacheError> {
+        // If none, create at default: (e.g. ~/.ocular/grpc_endpoints.toml)
+        let path: PathBuf = match file_path {
+            Some(path) => PathBuf::from(path),
+            None => {
+                let mut p = dirs::home_dir().unwrap();
+
+                p.push(DEFAULT_FILE_CACHE_DIR);
+                p.push(DEFAULT_FILE_CACHE_NAME);
+
+                p
+            }
+        };
+
+        dbg!(format!("Attempting to use path {:#?}", path));
+
+        // Verify path formatting
+        if path.extension().unwrap().to_str().unwrap() != "toml" {
+            return Err(CacheError::Initialization(String::from(
+                "Only toml files supported.",
+            )));
+        } else if path.is_dir() {
+            return Err(CacheError::Initialization(String::from(
+                "Path is a dir; must be a file.",
+            )));
+        }
+
+        // Create files/dirs based on override settings
+        // First just create dirs safely regardless of override settings
+        let save_path = path.parent().unwrap();
+
+        let dir_res = save_path.metadata();
+
+        // Create dir if doesn't exist
+        if dir_res.is_err() {
+            match std::fs::create_dir_all(save_path) {
+                Ok(_res) => _res,
+                Err(err) => return Err(CacheError::FileIO(err.to_string())),
+            };
+        }
+
+        #[cfg(unix)]
+        match std::fs::set_permissions(
+            save_path,
+            std::fs::Permissions::from_mode(FILE_CACHE_DIR_PERMISSIONS),
+        ) {
+            Ok(_res) => _res,
+            Err(err) => return Err(CacheError::FileIO(err.to_string())),
+        };
+
+        let mut endpoints = HashSet::new();
+
+        // Load endpoints if they exist
+        if std::path::Path::new(&path).exists() {
```

Comment:
pretty sure this can be `path.exists()`?

Reply:
Updated